File: 프로그래머스/lv2/68645. 삼각 달팽이/삼각 달팽이.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(int n) {
    vector<int> answer;
    
    vector<vector<int>> snail(n, vector<int>(n, 0)); //nxn 0으로 초기화
    int i=0, j=0, num=1; 
    
    while(1){
        bool end = true; 
        while(i+1<n && snail[i+1][j]==0){
            //아래로 갈 수 있을 때까지 내려가기 (하향)
            snail[i++][j] = num++;
            end = false; 
        }
        if(end) break; 
        while(j+1<n && snail[i][j+1]==0){
            snail[i][j++]=num++;
            end=false; 
        }
        if(end) break; 
        while(i-1>0 && j-1>0&&snail[i-1][j-1]==0){
            snail[i--][j--]=num++;
            end=false;
        }
        if(end) break;
    }
    snail[i][j]=num; 
    
    for(auto s: snail){
        for(auto si: s){
            if(si) answer.push_back(si);
        }
    }
    return answer;
}

File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.java
class Solution {
    public int[] solution(int n) {
        int[] answer = {};
        int size = n*(n+1)/2; 
        int[][] dal = new int [n][]; 
        for(int i=0; i<n; i++){
            dal[i] = new int[i+1]; 
        }
        int num = 1;
        int row = 0, col = 0; 
        
        int cnt = n; 
        while(cnt>0){
            while(row<n&&dal[row][col]==0){
                dal[row++][col] = num++; 
            }
            if(--cnt == 0) {break;}
            
            row--; 
            col++; 
            while(col<n&&dal[row][col]==0){
                dal[row][col++] = num++; 
            }
            if(--cnt == 0) {break;}

            col--; 
            row--; col--; 
            while(row>=0&&col>=0&&dal[row][col]==0){
                dal[row--][col--]=num++;
            }
            row++; col++;
            row++;
            if(--cnt == 0) {break;}
        }
        
        answer = new int[size];
        int idx = 0; 
        for(int i=0; i<n; i++){
            for(int j=0; j<dal[i].length; j++){
                answer[idx] = dal[i][j];
                idx++;
            }
        }
        return answer;
    }
}